### PR TITLE
RNN 循环神经网络系列 4: 注意力机制

### LINKED ISSUE #2253: RNN 循环神经网络系列 4: 注意力机制
* 原文链接：[RECURRENT NEURAL NETWORK (RNN) – PART 4: ATTENTIONAL INTERFACES](https://theneuralperspective.com/2016/11/20/recurrent-neural-network-rnn-part-4-attentional-interfaces/)
  * Markdown文件：[文件地址](https://github.com/xitu/gold-miner/blob/master/TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md)
  * PR 地址：#2280
  * 文章分类：**机器学习**
  ----
  * 翻译时间：6 天
  * 校对时间：2 天
  * 翻译积分：6 分
  * 校对积分：2 分
  * [积分有什么用](https://github.com/xitu/gold-miner/wiki)

### PR DESCRIPTION
**Pull request by TobiasLee**
译文翻译完成，resolve #2253
敬请斧正

### CODE REVIEW

**Comment by changkun** [comment]
校对认领

**Comment by fanyijihua** [comment]
@changkun 好的呢 :beer:

File: TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md [hunk 1]
```diff
@@ -14,144 +14,148 @@
...
+最后，总算到了注意力机制这一部分。我们已经知道了输入和输出，我们把一系列参数（时序列表、初始状态、注意力矩阵这些编码器的输出）交给了 `embedded_attention_decoder()` 函数，但在这其中究竟发生了什么？首先， 我们会创建一系列权重来对输入进行嵌入操作，我们把这些权重命名为 W_embedding。在通过输入生成解码器的输出之后，我们会开始一个循环函数，来决定将哪一部分输出交给下一个解码器作为输入。在训练过程中，我们通常不会把前一个解码器单元的输出传递给下一个，所以这里的循环函数是 None。而在推理期间，我们会这样做，所以这里的循环函数就会使用 `_extract_argmax_and_embed()`，它的用处就如它的名字所言（提取参数并且嵌入）。得到解码器单元的输出之后，让它和 softmax 后的权重矩阵相乘（output_projection），并将它的形状从 **[N, H]** 转换成 **[N, C]**，再使用同样 W_embedding 来替代经过嵌入操作的输出(**[N, H]**)，再将经过处理的输出作为下一个解码器单元的输入。
 
 ```
 # Loop function if using decoder outputs for next prediction
```
Comment (changkun) [inline TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: 删掉英文原文

File: TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md [hunk 2]
```diff
@@ -14,144 +14,148 @@
...
+得到动态 RNN 的输出和状态之后，我们就能够根据新的状态计算出新的上下文向量。cell 的输出和新的上下文向量再通过一个 MLP，最终就能得到我们的解码器输出。这些额外的 MLP 并没有在解码器的示意图中画出，但他们是我们得到输出必要的额外步骤。值得注意的是，cell 的输出和 attention_decoder 的输出的形状都是**[max_len, N, H]**。
 
-If we were doing inference, our loop function is no longer None but the _extract_argmax_and_append(). This takes in the previous decoder output and our new decoder input is just the previous output with softmax applied to it, following by reembedding. And after we do all the processing with the attention states, the prev is updated to be the newly predicted output.
+而当我们在进行推断的时候，循环函数不再是 None，而是 `_extract_argmax_and_append()`。这个函数会接收前一个解码器单元的输出，而我们新的解码器单元的输入就是先前的输出经过 softmax 之后的结果，接下来对它进行重嵌入操作。在利用注意力矩阵进行w完所有处理之后，将 prev 将被更新为新预测的输出。
 
 ```
 # Process decoder inputs one by one
```
Comment (changkun) [inline TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: 删除英文原文

File: TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md [hunk 3]
```diff
@@ -14,144 +14,148 @@
...
+# 挨个处理解码器的出入
```
Comment (changkun) [inline TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: 挨个处理解码器的出入
=>
依次处理解码器的输入

File: TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md [hunk 4]
```diff
@@ -14,144 +14,148 @@
...
 for i, inp in enumerate(decoder_inputs):
 
     if i &amp;gt; 0:
```
Comment (changkun) [inline TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: `&amp;gt;`
=>
`>`

File: TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md [hunk 5]
```diff
@@ -162,14 +166,16 @@ for i, inp in enumerate(decoder_inputs):
             inp = loop_function(prev, i)
 
     # Merge the input and attentions together
```
Comment (changkun) [inline TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: 删除

File: TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md [hunk 6]
```diff
@@ -162,14 +166,16 @@ for i, inp in enumerate(decoder_inputs):
...
+    # 把 输入和注意力向量合并
     input_size = inp.get_shape().with_rank(2)[1]
     x = tf.nn.rnn_cell._linear(
         args=[inp]+attns, output_size=input_size, bias=True)
 
-    # Decoder RNN
+    # 解码器 RNN
     cell_outputs, state = cell(x, state) # our stacked cell
 
     # Attention mechanism to get Cs
```
Comment (changkun) [inline TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: 删除

File: TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md [hunk 7]
```diff
@@ -162,14 +166,16 @@ for i, inp in enumerate(decoder_inputs):
...
+    # 通过注意力拿到 上下文向量
```
Comment (changkun) [inline TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: 删除空格

File: TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md [hunk 8]
```diff
@@ -218,32 +225,31 @@ else:
...
+### 带有 buckets 的模型:
 
-However, I find this method a bit crude and if we really wanted to process out the PADs, I would suggest using seq_lens to filter out the PADs first out of the encoder outputs OR once we compute the attention context vectors_i, we zero out the locations where the hidden state comes from a PAD for each sequence. This is a bit complicated and we won’t implement it here but buckets does serve as a decent solution for this extra noise.
+另外一种常见的模型会在 tf.nn.seq2seq 模型上添加 `_with_buckets()` 函数，这也是 Tensorflow 官方的 NMT [教程](https://www.tensorflow.org/versions/r0.11/tutorials/seq2seq/index.html)所使用的模型，这种 buckets 模型的优点在于缩短了注意力矩阵向量的长度。在先前的模型中，我们会把注意力向量应用在 max_len 长度的 hidden states 上。而在这里，我们只要对相关的一部分应用注意力向量，因为 PAD token 是完全可以被忽略的。我们可以选择对应的 buckets 使得句子中的 PAD token 尽可能的少。
```
Comment (changkun) [inline TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: 另外一种常见的模型会在 tf.nn.seq2seq 模型上添加 `_with_buckets()` 函数
=>
另外一种常见的附加结构是使用 `tf.nn.seq2seq.model_with_buckets()` 函数

**Comment by Brucexz** [comment]
校对认领

**Comment by fanyijihua** [comment]
@Brucexz 妥妥哒 :beers:

File: TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md [hunk 9]
```diff
@@ -14,144 +14,148 @@
...
 
 ![attention.png](https://github.com/ajarai/casual-digressions/blob/master/notes/images/rnn_attention/attention.png?raw=true)
 
-First we will take a look at the entire model and talk about some interest parts, then we will break down attention and then we will start the implementation with the same model as our encoder-decoder without attention model, which I have detailed **[here](https://theneuralperspective.com/2016/11/20/recurrent-neural-networks-rnn-part-3-encoder-decoder/)**. We will slowly introduce attention and we will also implement inferencing. **Note**: This is not going to be a state of the art model, especially since I wrote the data myself in couple minutes. This will purely be a post about understanding the architecture and you can use the implementation with your own larger data sources and you will achieve nice results.
+首先，让我们来一窥整个模型的架构并且讨论其中一些有趣的部分，然后我们会在先前实现的不带有注意力机制的编码器-解码器模型基础之上，添加注意力机制，先前模型的实现细节在[这里](https://theneuralperspective.com/2016/11/20/recurrent-neural-networks-rnn-part-3-encoder-decoder/)。**注意**：因为数据是我自己在分钟内草率地编写的，所以这个模型的性能可能并不是非常出色。这篇文章旨在帮助你理解带有注意力机制的模型，从而你能够运用到更大的数据集上，并且取得非常不错的结果。
```
Comment (Brucexz) [inline TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: 这篇文章旨在帮助你理解带有注意力机制的模型，从而你能够运用到更大的数据集上，并且取得非常不错的结果
=>
这篇文章仅仅是对于这一模型的理解，你可以把它运用到更大的数据集上，然后取得非常不错的结果

File: TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md [hunk 10]
```diff
@@ -14,144 +14,148 @@
...
 
-This diagram is a detailed version of the first diagram. We will start from the encoder and move up to the decoder outputs. We have out input data which has been padded and embedded. We feed these into our encoder RNN with a hidden state. The hidden state is initially all zeros but after our first input, it starts changing and holding valuable information. Just know that if you use an LSTM, we will also be passing a cell state c along with our hidden state h. For each input into the encoder we get the hidden state output which is passed for the next input but it is also used as the output for this cell for this input. We can call these h_1 to h_N and these some of our inputs for the attention model.
+这张图片是第一张图的更为具体的版本，包含了更多细节。让我们从编码器开始讲起，直到最后解码器的输出。首先，我们的输入数据是经过填充（Padding）和词嵌入（Embedding）处理的向量，我们将这些向量交给带有一系列 cell（上图中蓝色的 RNN 单元）的 RNN 网络，这些 cell 的输出称为隐藏状态（hidden state，上图中的h<sub>0</sub>，h<sub>1</sub>等)，它们被初始化为零，但在输入数据之后，这些隐藏状态会改变并且持有一些非常有价值的信息。如果你使用的是一个 LSTM 网络（RNN 的一种），我们会把 cell 的状态和它的输出一起向前传递给下一个 cell。对于每一个输入（上图中的 X<sub>0</sub>等），在每一个 cell 上我们都会得到一个隐藏状态的输出，这个输出也会作为下一个 cell 输入的一部分。我们把每个神经元的输出记作 h<sub>1</sub> 到 h<sub>N</sub>，这些输出将会成为我们注意力模型的输入。
```
[the remark this answers is not in the document]
Comment (Brucexz) [reply TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: 我们会把 cell 的状态 c 和隐藏状态 h 的一起传递给下一个 cell
=>
我们会把 cell 的状态 c 和隐藏状态 h 一起传递给下一个 cell

File: TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md [hunk 11]
```diff
@@ -14,144 +14,148 @@
...
+我为模型创建了一个很小的数据集：20 个英语和对应的西班牙语句子。这篇教程的重点是让你了解如何建立一个带有软注意力机制的编码器-解码器模型，来解决像机器翻译等的序列到序列问题。所以我写了关于我自己的 20 个英文句子，然后把他们翻译成对应的西班牙语，这就是我们的数据。
 
-First we separate the sentences into tokens and then convert the tokens into token ids. During this process we collect a vocabulary dict and a reverse vocabulary dict to convert back and forth between tokens and token ids. For our target language (spanish), we will add an extra EOS token. Then we will pad both source and target tokens to the max length (biggest sentence in the respective datasets). This is the data we feed into our model. We use the padded source inputs as is for the encoder, but we will do further additions to the target inputs in order to get our decoder inputs and outputs.
+首先，我们把这些句子变成一系列 token，再把 token 转换成对应的词汇 id。在这个处理过程中，我们会建立一个词汇词典，使我们能够从 token 和词汇 id 之间完成转换。对于我们的目标语言（西班牙语），我们会额外地添加一个 EOS 标识。接下来我们将对源语言和目标语言转换得来的一组 token 进行填充操作，将它们补齐至最大长度（数据集中最长的句子长度），这将成为最终我们要喂给我们模型的数据。我们把经过填充的源语言数据传给编码器，但我们还会对目标语言的输入做一些额外的操作以获得解码器的输入和输出。
```
Comment (Brucexz) [inline TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: 数据集中最长的句子长度
=>
分别是它们各自的数据集中的最长句子长度

Comment (Brucexz) [inline TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md]: 数据集中最长的句子长度
=>
它们各自所属的数据集中最长句子长度

**Comment by Brucexz** [comment]
译文质量极高，只有少数错字，标点错误
@TobiasLee @leviding 校对完毕

**Comment by Brucexz** [comment]
译文质量极高，只有少数错字
@TobiasLee @leviding 校对完毕

**Comment by leviding** [comment]
@TobiasLee 可以修改啦

**Comment by leviding** [comment]
已经 merge 啦~ 快快麻溜发布到掘金专栏然后给我发下链接，方便及时添加积分哟。

掘金翻译计划有自己的知乎专栏，你也可以投稿哈，推荐使用一个[好用的插件](https://chrome.google.com/webstore/detail/minimalist-markdown-edito/pghodfjepegmciihfhdipmimghiakcjf?hl=zh-CN)。
专栏地址：https://zhuanlan.zhihu.com/juejinfanyi

**Comment by TobiasLee** [comment]
https://juejin.im/post/59f72f61f265da432002871c  @leviding
